programming-languages.md

```diff
@@ -1,51 +1,82 @@
-develop people, and to be great at operating applications, we want to focus
-on a small number of choices of programming languages for core
-'backend development' tasks.
+We want to focus on using a small number of programming languages for
+core software development tasks.
 
-These are (alphabetically): Java, Python, Ruby. These are successfully being
-used by teams at the moment, and we are confident in how to host and operate
-applications written in them. New development should normally be carried out
-in one of them.
+This should make it easier for developers to:
 
-There will be sensible reasons to go outside this core group. For example:
+- move around the organisation
+- develop shared components
+- improve their personal development
+- master how they operate applications
 
-- when we're extending an existing codebase/ecosystem
-- when scripting in a particular environment
-- when we're experimenting during an alpha (with an expectation that it's replaced by something from the core set for beta)
+## Java, Python and Ruby
 
-The set of languages in the core group will change over time. This is likely
-to be through the adoption of a new language into the core group, and with
-an existing language being put on hold for new work.
+We've chosen these 3 languages because they are successfully used by
+teams at the moment, and we are confident in how to host and operate
+applications written in them. We'd advise you to carry out new
+development in one of them.
 
-## Languages in use not in the core group
+## JavaScript
 
-### Go
+This recommendation only applies to server-side JavaScript.
+The service manual has information on
+[using client-side JavaScript][manual_js].
 
-A small amount of use in GOV.UK. Core language for CloudFoundry used by the
-PaaS team.
+[manual_js]: https://www.gov.uk/service-manual/technology/using-progressive-enhancement
 
-### Node.js (JavaScript)
+The only way we're comfortable writing server-side JavaScript at the
+moment is with [Node.js][nodejs].
 
-In use in GOV.UK Pay and Performance Platform. Not as mature as other
-languages/frameworks in use, and lacking a critical mass of skills in
-GDS, so further adoption discouraged. Note: does not apply to in-browser
-JavaScript.
+[nodejs]: https://nodejs.org/
 
-### Scala
+Node.js is used by a few projects at GDS; GOV.UK Frontend, GOV.UK Pay and the Performance Platform.
 
-Used in the early days of GDS. GOV.UK Licensing is the only remaining
-application written in Scala. Not for use in new development.
+In the past we've had problems operating and maintaining products written in
+Node.js. We suspect that this is because:
+
+- It's younger and less mature than Java, Python and Ruby
+- We have fewer developers with Node.js skills than other tools we use
+
+Because both of these points should gradually improve over time, we think
+that Node.js is suitable for creating thin, client-facing applications
+that don't persist data.
```

Comment:
Looking at the skill sets of frontend developers (and also to a lesser extent backend developers) applying to GDS, NodeJS is an extremely common skill for people who have been working in the commercial web development environment in the past couple of years.  The quality and depth of the experience people have had varies massively, but it is clearly a technology in heavy adoption, particularly for rendering the web interface.

It is now rare to find frontend developers without some NodeJS experience.

I think the recommendation about persistence is probably a little more specific than I would word it, from what I've heard.  Perhaps replace this paragraph with something like:

> Since Node.js is now frequently used in the wider development ecosystem for building the web frontend and there are thus a lot of supporting tools and technologies, we expect this to gradually improve over time. It may therefore sometimes be appropriate to use this technology for rendering the web interface in our applications. It is hard to set a specific boundary for where Node.js should be used, but our current approach is that Node.js shouldn't be used to implement any significant "application logic". GOV.UK Pay are adopting this approach by using Node.js only to create thin, client-facing applications which do not directly persist data (eg, they may talk to an API for this); they transform an API to a web interface (and back again).  This divide matches reasonably well with the division between frontend and backend areas of specialism, and should allow the frontend to be developed with a reasonable level of decoupling from the backend applications.
>
> Teams that adopt Node.js must take steps to ensure that they have access to sufficient developers with Node.js experience.